lib/rubocop/cop/lint/non_deterministic_require_order.rb

```diff
@@ -0,0 +1,89 @@
+# frozen_string_literal: true
+
+module RuboCop
+  module Cop
+    module Lint
+      # `Dir[...]` and `Dir.glob(...)` do not make any guarantees about
+      # the order in which files are returned. The final order is
+      # determined by the operating system and file system.
+      # This means that using them in cases where the order matters,
+      # such as requiring files, can lead to intermittent failures
+      # that are hard to debug. To ensure this doesn't happen,
+      # always sort the list.
+      #
+      # @example
+      #
+      #   # bad
+      #   Dir["./lib/**/*.rb"].each do |file|
+      #     require file
+      #   end
+      #
+      #   # good
+      #   Dir["./lib/**/*.rb"].sort.each do |file|
+      #     require file
+      #   end
+      #
+      # @example
+      #
+      #   # bad
+      #   Dir.glob(Rails.root.join(__dir__, 'test', '*.rb')).each do |file|
+      #     require file
+      #   end
+      #
+      #   # good
+      #   Dir.glob(Rails.root.join(__dir__, 'test', '*.rb')).sort.each do |file|
+      #     require file
+      #   end
+      #
+      class NonDeterministicRequireOrder < Cop
+        MSG = 'Sort files before requiring them.'
+
+        def on_block(node)
+          return unless node.body
+          return unless unsorted_dir_loop?(node.send_node)
+
+          var_name = loop_variable(node.arguments)
+
+          return unless var_name && var_is_required?(node.body, var_name)
+
+          add_offense(node.send_node)
```

Comment:
The generated methods also take a block, which is yielded to in the case of a match. 🙂 

```ruby
loop_variable(node.arguments) do |var_name|
  return unless var_is_required?(node.body, var_name)

  add_offense(node.send_node)
end
```

Looks like you could also potentially combine the patterns into a single matcher.

Reply:
Ah, nice! I like the block syntax.

As for the single matcher, I like that all of the building blocks are fairly simple patterns. Merging them into one would make it harder to understand, I think.